File: wrappers/python/corporate_actions_registry/registry.py
"""
Main registry class for the Corporate Actions Registry.

Loads actions.json and provides fast lookup methods for actions by
ISIN, action ID, action type, and date range. The class is dependency-free
and uses the data models from .models for typed access.

Typical usage:
    from corporate_actions_registry import CorporateActionsRegistry

    registry = CorporateActionsRegistry("actions.json")
    aapl_actions = registry.by_isin("US0378331005")
    splits = registry.by_action_type("SPLIT")
"""

import json
import os
from typing import Any, Dict, List, Optional, Union
import copy

from .models import Action, RegistryMeta


class CorporateActionsRegistry:
    """
    A convenient interface to the Corporate Actions Registry data.

    The registry is loaded once into memory and then queried repeatedly
    without re-reading the file.

    Attributes:
        actions (List[Action]): List of all actions, parsed into Action objects.
        meta (RegistryMeta): Metadata about the registry file.
        _index_by_isin (Dict[str, List[Action]]): Precomputed index by ISIN.
        _index_by_type (Dict[str, List[Action]]): Precomputed index by action_type.
        _index_by_id (Dict[str, Action]): Precomputed index by action_id.
    """

    def __init__(
        self,
        actions_path: Optional[str] = None,
        actions_data: Optional[Dict[str, Any]] = None,
    ) -> None:
        """
        Initialize the registry from a file path or a pre-loaded dict.

        Either `actions_path` or `actions_data` must be provided.
        If both are provided, `actions_data` takes precedence.

        Args:
            actions_path: Path to actions.json.
            actions_data: Already loaded JSON data as a dict.

        Raises:
            FileNotFoundError: If actions_path does not exist.
            ValueError: If the JSON is invalid or missing required keys.
        """
        if actions_data is not None:
            raw_data = actions_data
        elif actions_path is not None:
            with open(actions_path, "r", encoding="utf-8-sig") as f:
                raw_data = json.load(f)
        else:
            raise ValueError("Either actions_path or actions_data must be provided.")

        if not isinstance(raw_data, dict) or "actions" not in raw_data:
            raise ValueError("Invalid actions data: expected dict with 'actions' key.")

        if not isinstance(raw_data["actions"], list):
            raise ValueError(
                "Invalid actions data: 'actions' must be a list, got "
                f"{type(raw_data['actions']).__name__}"
            )

        for i, item in enumerate(raw_data["actions"]):
            if not isinstance(item, dict):
                raise ValueError(
                    f"Invalid actions data: action at index {i} is not "
                    f"an object, got {type(item).__name__}"
                )

        self.meta = RegistryMeta.from_dict(raw_data.get("meta"))
        self.actions: List[Action] = [
            Action.from_dict(item) for item in raw_data["actions"]
        ]

        self._build_indexes()

    def _build_indexes(self) -> None:
        """Build in-memory indexes for fast lookups."""
        self._index_by_isin: Dict[str, List[Action]] = {}
        self._index_by_type: Dict[str, List[Action]] = {}
        self._index_by_id: Dict[str, Action] = {}

        for action in self.actions:
            if action.action_id:
                self._index_by_id[action.action_id] = action

            if action.isin:
                self._index_by_isin.setdefault(action.isin, []).append(action)

            if action.action_type:
                self._index_by_type.setdefault(action.action_type, []).append(action)

    # ------------------------------------------------------------------
    # Lookup methods
    # ------------------------------------------------------------------
    def by_isin(self, isin: str) -> List[Action]:
        """
        Return all actions for a given ISIN.

        Args:
            isin: The ISIN of the instrument (e.g., 'US0378331005').

        Returns:
            List of Action objects, possibly empty.
        """
        return [copy.deepcopy(a) for a in self._index_by_isin.get(isin, [])]
    
    def by_action_id(self, action_id: str) -> Optional[Action]:
        action = self._index_by_id.get(action_id)
        if action is None:
            return None
        # Return a copy so a caller cannot mutate the registry's internal
        # state by editing the returned object.
        return [copy.deepcopy(a) for a in self._index_by_type.get(action_type, [])]
    
    def by_action_type(self, action_type: str) -> List[Action]:
        """
        Return all actions of a given type (e.g., 'SPLIT', 'DIVIDEND').

        Args:
            action_type: The action type string.

        Returns:
            List of Action objects, possibly empty.
        """
        return list(self._index_by_type.get(action_type, []))

    _VALID_DATE_FIELDS = ("announcement", "ex_date", "record_date", "effective_date")

    def by_date_range(self, start_date=None, end_date=None, date_field="ex_date"):
        """Return actions whose `date_field` falls in [start_date, end_date].

        Bounds are inclusive. Pass None (or omit) for an open bound.
        Actions lacking `date_field` are skipped silently.

        Raises ValueError if `date_field` is not one of the four valid names.
        Results are sorted by the requested field, then by action_id.
        """
        if date_field not in self._VALID_DATE_FIELDS:
            raise ValueError(
                f"invalid date_field {date_field!r}; "
                f"expected one of {sorted(self._VALID_DATE_FIELDS)}"
            )
        result = []
        for action in self.actions:
            value = getattr(action.dates, date_field, None)
            if value is None:
                continue
            if start_date and value < start_date:
                continue
            if end_date and value > end_date:
                continue
            result.append(action)
        result.sort(key=lambda a: (
            getattr(a.dates, date_field, "") or "",
            a.action_id or "",
        ))
        return [copy.deepcopy(a) for a in result]
    
    def all_action_types(self) -> List[str]:
        """Return a sorted list of unique action types in the registry."""
        return sorted(self._index_by_type.keys())

    def count(self) -> int:
        """Return total number of actions."""
        return len(self.actions)

    def to_dict(self) -> Dict[str, Any]:
        """
        Serialize the entire registry back to a dictionary.

        Returns:
            Dict with 'meta' and 'actions' keys.
        """
        return {
            "meta": self.meta.to_dict(),
            "actions": [action.to_dict() for action in self.actions],
        }

    def save(self, path: str) -> None:
        """
        Save the registry to a JSON file.

        Args:
            path: Destination file path.
        """
        with open(path, "w", encoding="utf-8") as f:
            json.dump(self.to_dict(), f, indent=2, ensure_ascii=False)

    def __len__(self) -> int:
        """Return number of actions."""
        return len(self.actions)

    def __repr__(self) -> str:
        return f"<CorporateActionsRegistry actions={len(self.actions)}>"

File: wrappers/go/registry/registry.go
// Package registry provides a typed interface to the Corporate Actions Registry.
//
// It loads actions.json and builds in-memory indexes for fast lookups by
// ISIN, action ID, and action type, along with date range filtering.
//
// The behaviour of every method in this package is specified by
// docs/wrapper_contract.md. When the code disagrees with that document,
// the code is wrong.
package registry

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"sort"
)

// ErrInvalidDateField is returned by ByDateRange when the requested date
// field name is not one of the four valid values.
var ErrInvalidDateField = errors.New("invalid date_field")

// validDateFields is the closed set of date field names accepted by
// ByDateRange. Kept as a map for O(1) lookup.
var validDateFields = map[string]bool{
	"announcement":   true,
	"ex_date":        true,
	"record_date":    true,
	"effective_date": true,
}

// Dates contains the announcement, ex, record, and effective dates of a
// corporate action. All fields are optional.
type Dates struct {
	Announcement  *string `json:"announcement,omitempty"`
	ExDate        *string `json:"ex_date,omitempty"`
	RecordDate    *string `json:"record_date,omitempty"`
	EffectiveDate *string `json:"effective_date,omitempty"`
}

// Provenance contains source and verification information for an action.
type Provenance struct {
	Source             *string `json:"source,omitempty"`
	SourceURL          *string `json:"source_url,omitempty"`
	VerificationSource *string `json:"verification_source,omitempty"`
	VerificationURL    *string `json:"verification_url,omitempty"`
}

// Impact contains financial impact multipliers for backtesting.
type Impact struct {
	PriceMultiplier *float64 `json:"price_multiplier,omitempty"`
	ShareMultiplier *float64 `json:"share_multiplier,omitempty"`
	CashAdjustment  *float64 `json:"cash_adjustment,omitempty"`
}

// Action represents a single corporate action entry.
type Action struct {
	ISIN       *string     `json:"isin,omitempty"`
	ActionID   *string     `json:"action_id,omitempty"`
	ActionType *string     `json:"action_type,omitempty"`
	Ratio      *string     `json:"ratio,omitempty"`
	Amount     *float64    `json:"amount,omitempty"`
	Currency   *string     `json:"currency,omitempty"`
	Dates      *Dates      `json:"dates,omitempty"`
	Status     *string     `json:"status,omitempty"`
	Provenance *Provenance `json:"provenance,omitempty"`
	Impact     *Impact     `json:"impact,omitempty"`
}

// Meta contains metadata about the registry file.
type Meta struct {
	Version     *string `json:"version,omitempty"`
	GeneratedAt *string `json:"generated_at,omitempty"`
	Source      *string `json:"source,omitempty"`
	Notes       *string `json:"notes,omitempty"`
}

// Registry is the main container for all corporate actions.
//
// A Registry is immutable after construction. All lookup methods return
// fresh slices or freshly-allocated pointers, so a caller cannot corrupt
// the registry's internal state by mutating a returned value.
type Registry struct {
	meta    Meta
	actions []Action

	// Indexes. Each maps a key to one or more positions in `actions`.
	// Positions are stable for the lifetime of the Registry.
	indexISIN map[string][]int
	indexType map[string][]int
	indexID   map[string]int
}

// LoadRegistry loads the registry from a JSON file.
//
// A UTF-8 BOM at the start of the file is ignored. Any other leading
// bytes are passed through to json.Unmarshal.
func LoadRegistry(path string) (*Registry, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("reading registry file: %w", err)
	}
	return FromJSON(data)
}

// FromJSON parses the registry from raw JSON bytes.
//
// The document must be an object with an "actions" array. The "meta"
// object is optional. Each entry in "actions" must itself be a JSON
// object; a string, number, boolean, null, or array entry is a
// structural error, reported with its index.
//
// A UTF-8 BOM prefix is stripped before parsing.
func FromJSON(data []byte) (*Registry, error) {
	data = bytes.TrimPrefix(data, []byte{0xEF, 0xBB, 0xBF})

	// Decode the actions array into raw messages first so each entry can
	// be inspected before it is unmarshaled into an Action. This is the
	// only way to distinguish a JSON object from a JSON array, string,
	// number, boolean, or null inside the array.
	var raw struct {
		Meta    *Meta              `json:"meta"`
		Actions *[]json.RawMessage `json:"actions"`
	}
	if err := json.Unmarshal(data, &raw); err != nil {
		return nil, fmt.Errorf("parsing registry JSON: %w", err)
	}
	if raw.Actions == nil {
		return nil, fmt.Errorf("parsing registry JSON: missing 'actions' key")
	}

	actions := make([]Action, len(*raw.Actions))
	for i, item := range *raw.Actions {
		trimmed := bytes.TrimSpace(item)
		if len(trimmed) == 0 || trimmed[0] != '{' {
			return nil, fmt.Errorf(
				"parsing registry JSON: action at index %d is not an object",
				i,
			)
		}
		if err := json.Unmarshal(item, &actions[i]); err != nil {
			return nil, fmt.Errorf(
				"parsing registry JSON: action at index %d: %w",
				i, err,
			)
		}
	}

	meta := Meta{}
	if raw.Meta != nil {
		meta = *raw.Meta
	}

	r := &Registry{
		meta:      meta,
		actions:   actions,
		indexISIN: make(map[string][]int),
		indexType: make(map[string][]int),
		indexID:   make(map[string]int),
	}

	for i, action := range r.actions {
		if action.ISIN != nil {
			r.indexISIN[*action.ISIN] = append(r.indexISIN[*action.ISIN], i)
		}
		if action.ActionType != nil {
			r.indexType[*action.ActionType] = append(r.indexType[*action.ActionType], i)
		}
		if action.ActionID != nil {
			// Last one wins. Duplicate action_id values are a data error
			// caught by the validator's uniqueness layer; the loader does
			// not check.
			r.indexID[*action.ActionID] = i
		}
	}

	return r, nil
}

// ByISIN returns all actions for the given ISIN.
//
// Returns an empty slice when no action matches. Matching is exact and
// case-sensitive.
//
// The returned slice is fresh: mutating it does not affect the registry.
func (r *Registry) ByISIN(isin string) []Action {
	indices, ok := r.indexISIN[isin]
	if !ok {
		return []Action{}
	}
	result := make([]Action, len(indices))
	for i, idx := range indices {
		result[i] = r.actions[idx]
	}
	return result
}

// ByActionID returns a single action by its unique action_id, or nil if
// no action matches. Matching is exact and case-sensitive.
//
// The returned pointer refers to a freshly-allocated copy; mutating the
// pointed-to Action does not affect the registry.
func (r *Registry) ByActionID(actionID string) *Action {
	idx, ok := r.indexID[actionID]
	if !ok {
		return nil
	}
	src := r.actions[idx]
	// Copy the top-level struct; this copies the pointers to the nested
	// structs. Deep-copy those so a caller who mutates a nested field
	// does not reach the registry's internal state. Matches the deep-copy
	// contract in docs/wrapper_contract.md section 4.2.
	dst := src
	if src.Dates != nil {
		d := *src.Dates
		dst.Dates = &d
	}
	if src.Provenance != nil {
		p := *src.Provenance
		dst.Provenance = &p
	}
	if src.Impact != nil {
		i := *src.Impact
		dst.Impact = &i
	}
	return &dst
}

// ByActionType returns all actions of a given type (e.g., "SPLIT",
// "DIVIDEND"). Matching is exact and case-sensitive.
//
// The returned slice is fresh: mutating it does not affect the registry.
func (r *Registry) ByActionType(actionType string) []Action {
	indices, ok := r.indexType[actionType]
	if !ok {
		return []Action{}
	}
	result := make([]Action, len(indices))
	for i, idx := range indices {
		result[i] = r.actions[idx]
	}
	return result
}

// ByDateRange returns actions whose dateField falls in the inclusive
// range [startDate, endDate].
//
// dateField must be one of "announcement", "ex_date", "record_date", or
// "effective_date". Any other value returns ErrInvalidDateField.
//
// An empty startDate or endDate means "no bound on that side". Actions
// whose dateField is absent (or whose dates object is nil) are silently
// skipped. Dates are compared as ISO-8601 strings, so lexicographic
// comparison is chronological.
//
// The result is sorted by (dateField, action_id), ascending.
func (r *Registry) ByDateRange(startDate, endDate, dateField string) ([]Action, error) {
	if !validDateFields[dateField] {
		return nil, fmt.Errorf("%w: %s", ErrInvalidDateField, dateField)
	}

	dateValue := func(a Action) string {
		if a.Dates == nil {
			return ""
		}
		switch dateField {
		case "announcement":
			if a.Dates.Announcement != nil {
				return *a.Dates.Announcement
			}
		case "ex_date":
			if a.Dates.ExDate != nil {
				return *a.Dates.ExDate
			}
		case "record_date":
			if a.Dates.RecordDate != nil {
				return *a.Dates.RecordDate
			}
		case "effective_date":
			if a.Dates.EffectiveDate != nil {
				return *a.Dates.EffectiveDate
			}
		}
		return ""
	}

	result := make([]Action, 0)
	for _, action := range r.actions {
		v := dateValue(action)
		if v == "" {
			continue
		}
		if startDate != "" && v < startDate {
			continue
		}
		if endDate != "" && v > endDate {
			continue
		}
		result = append(result, action)
	}

	sort.Slice(result, func(i, j int) bool {
		vi, vj := dateValue(result[i]), dateValue(result[j])
		if vi != vj {
			return vi < vj
		}
		var ai, aj string
		if result[i].ActionID != nil {
			ai = *result[i].ActionID
		}
		if result[j].ActionID != nil {
			aj = *result[j].ActionID
		}
		return ai < aj
	})

	return result, nil
}

// AllActionTypes returns a sorted list of all unique action types present
// in the registry. Reserved types that are not present in the data do not
// appear. The result is empty for an empty registry.
func (r *Registry) AllActionTypes() []string {
	types := make([]string, 0, len(r.indexType))
	for t := range r.indexType {
		types = append(types, t)
	}
	sort.Strings(types)
	return types
}

// Count returns the total number of actions.
func (r *Registry) Count() int {
	return len(r.actions)
}

// Meta returns a pointer to the registry metadata. The metadata is a
// value type, so the pointer refers to a copy; mutating the fields of
// the returned Meta through their own pointers still reaches the shared
// strings, matching the behaviour of the Python and JavaScript wrappers.
func (r *Registry) Meta() *Meta {
	m := r.meta
	return &m
}

// ToJSON marshals the registry back to a JSON representation with
// exactly two top-level keys: "meta" and "actions".
//
// The output is indented with two spaces and carries no BOM.
func (r *Registry) ToJSON() ([]byte, error) {
	return json.MarshalIndent(struct {
		Meta    Meta     `json:"meta"`
		Actions []Action `json:"actions"`
	}{
		Meta:    r.meta,
		Actions: r.actions,
	}, "", "  ")
}

// Save writes the registry to a file.
//
// The output is UTF-8 JSON without a BOM. The file is created with mode
// 0644. Any existing file at the same path is overwritten.
func (r *Registry) Save(path string) error {
	data, err := r.ToJSON()
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0644)
}
